usgs tests are flaky

This test is currently failing:

https://github.com/oceanmodeling/searvey/blob/2a510e953a2799047553f1c253d8990141ab8906/tests/usgs_test.py#L90-L105

with:
```
=============================================================================================================== FAILURES ===============================================================================================================
__________________________________________________________________________________________ test_get_usgs_station_data[truncate_seconds=False] __________________________________________________________________________________________
tests/usgs_test.py:105: in test_get_usgs_station_data
    assert len(df) == no_records
E   assert 18 == 11
```

The dataframe we get is:
``` python
df = usgs.get_usgs_station_data(
    usgs_code="15484000",
    endtime=datetime.date(2022, 10, 1),
    period=1,
    truncate_seconds=False,
)
```
``` 
                                                                  value qualifier unit                                               name
site_no  datetime                  code  option                                                                                          
15484000 2022-09-30 08:00:00+00:00 00065                           7.32         A   ft                                  Gage height, feet
         2022-09-30 16:30:00+00:00 00065                           7.31         A   ft                                  Gage height, feet
         2022-10-01 00:00:00+00:00 00065                            7.3         A   ft                                  Gage height, feet
         2022-10-01 02:30:00+00:00 00065                           7.28         A   ft                                  Gage height, feet
         2022-10-01 04:15:00+00:00 00065                           7.29         A   ft                                  Gage height, feet
         2022-10-01 19:00:00+00:00 00065                           7.27         A   ft                                  Gage height, feet
         2022-10-02 03:15:00+00:00 00065                           7.26         A   ft                                  Gage height, feet
         2022-09-30 08:00:00+00:00 00065 ak dot&pf bridge datum  639.04         P   ft                                  Gage height, feet
         2022-09-30 16:30:00+00:00 00065 ak dot&pf bridge datum  639.03         P   ft                                  Gage height, feet
         2022-10-01 00:00:00+00:00 00065 ak dot&pf bridge datum  639.02         P   ft                                  Gage height, feet
         2022-10-01 02:30:00+00:00 00065 ak dot&pf bridge datum   639.0         P   ft                                  Gage height, feet
         2022-10-01 04:15:00+00:00 00065 ak dot&pf bridge datum  639.01         P   ft                                  Gage height, feet
         2022-10-01 19:00:00+00:00 00065 ak dot&pf bridge datum  638.99         P   ft                                  Gage height, feet
         2022-10-02 03:15:00+00:00 00065 ak dot&pf bridge datum  638.98         P   ft                                  Gage height, feet
         2022-09-30 08:25:00+00:00 72414 ak dot&pf bridge datum   632.3         P   ft  Streambed elevation at measurement point, abov...
         2022-09-30 08:55:00+00:00 72414 ak dot&pf bridge datum  632.21         P   ft  Streambed elevation at measurement point, abov...
         2022-09-30 12:55:00+00:00 72414 ak dot&pf bridge datum  632.12         P   ft  Streambed elevation at measurement point, abov...
         2022-09-30 14:55:00+00:00 72414 ak dot&pf bridge datum  632.38         P   ft  Streambed elevation at measurement point, abov...
[18 rows x 4 columns]
```

-------

I guess that USGS started to return the additional 7 rows recently. Judging from the output, the extra rows that cause the issue are the first 7 ones, i.e. the ones without an `option`.

Is it normal for USGS to add data like this?

Anyhow, this specific test does not seem to be testing what it is supposed to test (i.e. truncate_seconds), because the timestamps don't contain any seconds anyhow. We should probably choose a different combination of `usgs_station` and/or time period.  Furthermore, it will probably be more robust if we make the assertion like this:

``` python
df = usgs.get_usgs_station_data(..., truncate_seconds=False)
df_trunc = usgs.get_usgs_station_data(..., truncate_seconds=True)
assert len(df) > len(df_trunc)

seconds_trunc = df.index.to_frame().datetime.dt.second
assert (seconds_trunc == 0).all()
assert df_with
```

pinging @SorooshMani-NOAA 

PS. There are more tests that are failing, probably due to USGS returning extra data but I didn't dig in deeper in those.
```
FAILED tests/usgs_test.py::test_get_usgs_station_data[truncate_seconds=True] - assert 18 == 11
FAILED tests/usgs_test.py::test_get_usgs_station_data[truncate_seconds=False] - assert 18 == 11
FAILED tests/usgs_test.py::test_get_usgs_data[truncate_seconds=True] - AssertionError: assert 226 == 222
FAILED tests/usgs_test.py::test_get_usgs_data[truncate_seconds=False] - AssertionError: assert 226 == 222
FAILED tests/usgs_test.py::test_get_usgs_station_data_by_string_enddate - assert 30 == 17
```

I don't understand what USGS is doing...
```
Python 3.9 on Ubuntu: All tests passed
Python 3.10 on Macos:  FAILED tests/usgs_test.py::test_get_usgs_stations - assert 34945 > 35000
```
https://github.com/pmav99/searvey/actions/runs/5087897730/jobs/9143712140#step:12:259

--- a/searvey/usgs.py
+++ b/searvey/usgs.py
@@ -365,7 +365,7 @@ def get_usgs_data(
             continue
 
         ds = _get_dataset_from_query_results(result.result, usgs_metadata)
-        if len(ds) > 0:
+        if any((ds.count(ds.dims) > 0)[v] for v in ds.data_vars):
             datasets.append(ds)
 
     # in order to keep memory consumption low, let's group the datasets

--- a/searvey/usgs.py
+++ b/searvey/usgs.py
@@ -15,7 +15,6 @@ import datetime
 import functools
 import importlib.metadata
 import logging
-import warnings
 from itertools import product
 from typing import Any
 from typing import Dict
@@ -221,9 +220,7 @@ def get_usgs_stations(
     return usgs_stations
 
 
-def normalize_usgs_station_data(df: pd.DataFrame, truncate_seconds: bool) -> pd.DataFrame:
-    # TODO: Does truncate seconds make sense for USGS?
-
+def normalize_usgs_station_data(df: pd.DataFrame) -> pd.DataFrame:
     if df.empty:
         return df
 
@@ -258,16 +255,6 @@ def normalize_usgs_station_data(df: pd.DataFrame, truncate_seconds: bool) -> pd.
     df["unit"] = df_parm.parm_unit[df.code.values].values
     df["name"] = df_parm.parm_nm[df.code.values].values
 
-    if truncate_seconds:
-        # Truncate seconds from timestamps: https://stackoverflow.com/a/28783971/592289
-        # WARNING: This can potentially lead to duplicates!
-        df = df.assign(datetime=df.datetime.dt.floor("min"))
-        if df.duplicated(subset=list(USGS_DATA_MULTIIDX)).any():
-            # There are duplicates. Keep the first datapoint per minute.
-            msg = "Duplicate timestamps have been detected after the truncation of seconds. Keeping the first datapoint per minute"
-            warnings.warn(msg)
-            df = df.drop_duplicates(subset=list(USGS_DATA_MULTIIDX)).reset_index(drop=True)
-
     df = df.set_index(list(USGS_DATA_MULTIIDX))
 
     return df
@@ -277,7 +264,6 @@ def get_usgs_station_data(
     usgs_code: str,
     endtime: DateLike = TODAY,
     period: float = 30,
-    truncate_seconds: bool = True,
     rate_limit: Optional[RateLimit] = RateLimit(),
 ) -> pd.DataFrame:
     """Retrieve the TimeSeries of a single USGS station.
@@ -285,7 +271,6 @@ def get_usgs_station_data(
     :param usgs_code: USGS station code a.k.a. "site number"
     :param endtime: The end date for the measurement data for fetch. Defaults to `datetime.date.today()`
     :param period: Number of date for which to fetch station data
-    :param truncate_seconds: If ``True`` then timestamps are truncated to minutes (seconds are dropped)
     :param rate_limit: The default rate limit is 5 requests/second.
     :return: ``pandas.DataFrame`` with the a single station measurements
     """
@@ -297,19 +282,20 @@ def get_usgs_station_data(
     endtime = resolve_date(endtime)
     starttime = endtime - datetime.timedelta(days=period)
     df_iv, _ = nwis.get_iv(sites=[usgs_code], start=starttime.isoformat(), end=endtime.isoformat())
-    df_iv = normalize_usgs_station_data(df=df_iv, truncate_seconds=truncate_seconds)
+    df_iv = normalize_usgs_station_data(df=df_iv)
     return df_iv
 
 
 def _get_dataset_from_query_results(
-    query_result: Tuple[pd.DataFrame, Metadata], usgs_metadata: pd.DataFrame, truncate_seconds: bool
+    query_result: Tuple[pd.DataFrame, Metadata],
+    usgs_metadata: pd.DataFrame,
 ) -> xr.Dataset:
     df_iv, _ = query_result
     if len(df_iv) == 0:
         return xr.Dataset()
 
     df_iv = df_iv.reset_index()
-    df_iv = normalize_usgs_station_data(df=df_iv, truncate_seconds=truncate_seconds)
+    df_iv = normalize_usgs_station_data(df=df_iv)
     st_meta = (
         usgs_metadata.set_index("site_no")
         .loc[df_iv.reset_index().site_no.unique()]
@@ -334,40 +320,15 @@ def get_usgs_data(
     usgs_metadata: pd.DataFrame,
     endtime: DateLike = TODAY,
     period: float = 1,  # one day
-    truncate_seconds: bool = True,
     rate_limit: RateLimit = RateLimit(),
     disable_progress_bar: bool = False,
 ) -> xr.Dataset:
     """
     Return the data of the stations specified in ``usgs_metadata`` as an ``xr.Dataset``.
 
-    ``truncate_seconds`` needs some explaining. USGS has more than 1000 stations.
-    When you retrieve data from all (or at least most of) these stations, you
-    end up with thousands of timestamps that only contain a single datapoint.
-    This means that the returned ``xr.Dataset`` will contain a huge number of ``NaN``
-    which means that you will need a huge amount of RAM.
-
-    In order to reduce the amount of the required RAM we reduce the number of timestamps
-    by truncating the seconds. This is how this works:
-
-        2014-01-03 14:53:02 -> 2014-01-03 14:53:00
-        2014-01-03 14:53:32 -> 2014-01-03 14:53:00
-        2014-01-03 14:53:48 -> 2014-01-03 14:53:00
-        2014-01-03 14:54:09 -> 2014-01-03 14:54:00
-        2014-01-03 14:54:48 -> 2014-01-03 14:54:00
-
-    Nevertheless this approach has a downside. If a station returns multiple datapoints
-    within the same minute, then we end up with duplicate timestamps. When this happens
-    we only keep the first datapoint and drop the subsequent ones. So potentially you
-    may not retrieve all of the available data.
-
-    If you don't want this behavior, set ``truncate_seconds`` to ``False`` and you
-    will retrieve the full data.
-
     :param usgs_metadata: A ``pd.DataFrame`` returned by ``get_usgs_stations``.
     :param endtime: The date of the "end" of the data. Defaults to `datetime.date.today()`
     :param period: The number of days to be requested. USGS does not support values greater than 30
-    :param truncate_seconds: If ``True`` then timestamps are truncated to minutes (seconds are dropped)
     :param rate_limit: The default rate limit is 5 requests/second.
     :param disable_progress_bar: If ``True`` then the progress bar is not displayed.
     :return: ``xr.Dataset`` of station measurements
@@ -403,7 +364,7 @@ def get_usgs_data(
             # in `data_retrieval` due to non-existant data
             continue
 
-        ds = _get_dataset_from_query_results(result.result, usgs_metadata, truncate_seconds)
+        ds = _get_dataset_from_query_results(result.result, usgs_metadata)
         if len(ds) > 0:
             datasets.append(ds)
 

--- a/searvey/usgs.py
+++ b/searvey/usgs.py
@@ -305,6 +305,9 @@ def _get_dataset_from_query_results(
     query_result: Tuple[pd.DataFrame, Metadata], usgs_metadata: pd.DataFrame, truncate_seconds: bool
 ) -> xr.Dataset:
     df_iv, _ = query_result
+    if len(df_iv) == 0:
+        return xr.Dataset()
+
     df_iv = df_iv.reset_index()
     df_iv = normalize_usgs_station_data(df=df_iv, truncate_seconds=truncate_seconds)
     st_meta = (
@@ -401,7 +404,8 @@ def get_usgs_data(
             continue
 
         ds = _get_dataset_from_query_results(result.result, usgs_metadata, truncate_seconds)
-        datasets.append(ds)
+        if len(ds) > 0:
+            datasets.append(ds)
 
     # in order to keep memory consumption low, let's group the datasets
     # and merge them in batches

--- a/tests/usgs_test.py
+++ b/tests/usgs_test.py
@@ -146,10 +146,10 @@ def test_get_usgs_data():
 
     # Check that actual data has been retrieved
     assert ds.isel(site_no=0, code=0).value.size > 0
-    assert not np.isnan(ds.isel(site_no=0, code=0).value.values.astype(float)).all()
+    assert ds.isel(site_no=0, code=0).value.notnull().any()
 
     assert ds.isel(site_no=1, code=0).value.size > 0
-    assert not np.isnan(ds.isel(site_no=1, code=0).value.values.astype(float)).all()
+    assert ds.isel(site_no=1, code=0).value.notnull().any()
 
     assert all(col in ds.dims for col in ["site_no", "datetime", "code", "option"])
     assert all(col in ds.variables for col in ["value", "qualifier", "lon", "lat", "unit", "name"])

--- a/tests/usgs_test.py
+++ b/tests/usgs_test.py
@@ -1,6 +1,7 @@
 import datetime
 
 import geopandas as gpd
+import numpy as np
 import pandas as pd
 import pytest
 import xarray as xr
@@ -74,7 +75,6 @@ def test_get_usgs_stations():
     stations = usgs.get_usgs_stations()
     assert isinstance(stations, pd.DataFrame)
     assert isinstance(stations, gpd.GeoDataFrame)
-    assert 40000 > len(stations) > 30000
     # check that the DataFrame has the right columns
     assert set(stations.columns).issuperset(usgs.USGS_STATIONS_COLUMN_NAMES)
     # Check that the parsing of the dates has been done correctly
@@ -88,13 +88,24 @@ def test_usgs_metadata_has_state_info():
 
 
 def test_get_usgs_station_data():
-    """Truncate_seconds=False returns more datapoints compared to `=True`"""
+    # TODO: Parameterize the test for station, date and offset
     df = usgs.get_usgs_station_data(
         usgs_code="15484000",
         endtime=datetime.date(2022, 10, 1),
         period=1,
     )
-    assert len(df) == 18
+    assert all(col in df.columns for col in ["value", "qualifier", "unit", "name"])
+    assert isinstance(df.index, pd.MultiIndex)
+    assert all(col in df.index.names for col in ["site_no", "datetime", "code", "option"])
+
+    station_utc_offset = "-08:00"
+    times = df.index.get_level_values("datetime")
+    assert all(
+        np.logical_and(
+            f"2022-09-30 00:00:00{station_utc_offset}" <= times,
+            times <= f"2022-10-02 00:00:00{station_utc_offset}",
+        )
+    )
 
 
 _USGS_METADATA_MINIMAL = pd.DataFrame.from_dict(
@@ -125,7 +136,6 @@ def test_get_usgs_data():
         period=2,
     )
     assert isinstance(ds, xr.Dataset)
-    assert len(ds.datetime) == 226
     assert len(ds.site_no) == len(usgs_metadata)
     # Check that usgs_code, lon, lat, country and location are not modified
     assert set(ds.site_no.values).issubset(usgs_metadata.site_no.values)
@@ -133,10 +143,16 @@ def test_get_usgs_data():
     assert set(ds.lat.values).issubset(usgs_metadata.dec_lat_va.values)
     # assert set(ds.country.values).issubset(usgs_metadata.country.values)
     # assert set(ds.location.values).issubset(usgs_metadata.location.values)
+
     # Check that actual data has been retrieved
-    code = "00065"
-    assert ds.sel(site_no="15056500", code=code).value.mean() == pytest.approx(125.30, rel=1e-3)
-    assert ds.sel(site_no="15484000", code=code).value.mean() == pytest.approx(323.24, rel=1e-3)
+    assert ds.isel(site_no=0, code=0).value.size > 0
+    assert not np.isnan(ds.isel(site_no=0, code=0).value.values.astype(float)).all()
+
+    assert ds.isel(site_no=1, code=0).value.size > 0
+    assert not np.isnan(ds.isel(site_no=1, code=0).value.values.astype(float)).all()
+
+    assert all(col in ds.dims for col in ["site_no", "datetime", "code", "option"])
+    assert all(col in ds.variables for col in ["value", "qualifier", "lon", "lat", "unit", "name"])
 
 
 def test_get_usgs_station_data_by_string_enddate():
@@ -146,13 +162,19 @@ def test_get_usgs_station_data_by_string_enddate():
         period=2,
     )
 
-    assert len(df) == 30
+    station_utc_offset = -8
 
-    dates = pd.to_datetime(df.index.get_level_values("datetime").to_series().dt.date.unique())
-    assert len(dates) == 4
-    assert dates.day.tolist() == [29, 30, 1, 2]  # Why does it get the 2nd? timezone?
-    assert dates.month.tolist() == [9, 9, 10, 10]
-    assert dates.year.tolist() == [2022, 2022, 2022, 2022]
+    dates = pd.to_datetime(
+        df.index.get_level_values("datetime")
+        .to_series()
+        .reset_index(drop=True)
+        .dt.tz_convert(datetime.timezone(datetime.timedelta(hours=station_utc_offset)))
+        .dt.date.unique()
+    )
+    assert len(dates) == 3
+    assert dates.day.tolist() == [29, 30, 1]
+    assert dates.month.tolist() == [9, 9, 10]
+    assert dates.year.tolist() == [2022, 2022, 2022]
 
 
 def test_normalize_empty_stations_df():

--- a/tests/usgs_test.py
+++ b/tests/usgs_test.py
@@ -209,4 +209,4 @@ def test_request_nonexistant_data():
         endtime="2015-07-04",
         period=1,
     )
-    assert len(ds) == 0
+    assert all((ds.count(ds.dims) == 0)[v] for v in ds.data_vars)

--- a/tests/usgs_test.py
+++ b/tests/usgs_test.py
@@ -87,22 +87,14 @@ def test_usgs_metadata_has_state_info():
     assert "us_state" in stations.columns
 
 
-@pytest.mark.parametrize(
-    "truncate_seconds,no_records",
-    [
-        pytest.param(True, 18, id="truncate_seconds=True"),
-        pytest.param(False, 18, id="truncate_seconds=False"),
-    ],
-)
-def test_get_usgs_station_data(truncate_seconds, no_records):
+def test_get_usgs_station_data():
     """Truncate_seconds=False returns more datapoints compared to `=True`"""
     df = usgs.get_usgs_station_data(
         usgs_code="15484000",
         endtime=datetime.date(2022, 10, 1),
         period=1,
-        truncate_seconds=truncate_seconds,
     )
-    assert len(df) == no_records
+    assert len(df) == 18
 
 
 _USGS_METADATA_MINIMAL = pd.DataFrame.from_dict(
@@ -122,14 +114,7 @@ _USGS_METADATA_MINIMAL = pd.DataFrame.from_dict(
 )
 
 
-@pytest.mark.parametrize(
-    "truncate_seconds",
-    [
-        pytest.param(True, id="truncate_seconds=True"),
-        pytest.param(False, id="truncate_seconds=False"),
-    ],
-)
-def test_get_usgs_data(truncate_seconds):
+def test_get_usgs_data():
     # in order to speed up the execution time of the test,
     # we don't retrieve the USGS metadata from the internet,
     # but we use a hardcoded dict instead
@@ -138,10 +123,8 @@ def test_get_usgs_data(truncate_seconds):
         usgs_metadata=usgs_metadata,
         endtime="2022-09-29",
         period=2,
-        truncate_seconds=truncate_seconds,
     )
     assert isinstance(ds, xr.Dataset)
-    # TODO: Check if truncate_seconds does work
     assert len(ds.datetime) == 226
     assert len(ds.site_no) == len(usgs_metadata)
     # Check that usgs_code, lon, lat, country and location are not modified
@@ -161,7 +144,6 @@ def test_get_usgs_station_data_by_string_enddate():
         usgs_code="15484000",
         endtime="2022-10-01",
         period=2,
-        truncate_seconds=False,
     )
 
     assert len(df) == 30
@@ -183,7 +165,7 @@ def test_normalize_empty_stations_df():
 
 def test_normalize_empty_data_df():
     df = pd.DataFrame()
-    df2 = usgs.normalize_usgs_station_data(df, truncate_seconds=False)
+    df2 = usgs.normalize_usgs_station_data(df)
 
     assert df2.empty
     assert isinstance(df2, pd.DataFrame)

--- a/tests/usgs_test.py
+++ b/tests/usgs_test.py
@@ -191,3 +191,22 @@ def test_normalize_empty_data_df():
 
     assert df2.empty
     assert isinstance(df2, pd.DataFrame)
+
+
+def test_request_nonexistant_data():
+    sta = usgs.get_usgs_stations()
+    sta = sta[
+        sta.site_no.isin(
+            [
+                "373707086300703",
+                "373707086300801",
+                "373707086300802",
+            ]
+        )
+    ]
+    ds = usgs.get_usgs_data(
+        usgs_metadata=sta,
+        endtime="2015-07-04",
+        period=1,
+    )
+    assert len(ds) == 0

